Update to handle API reading and optimize sharded writing and indexing

iliat opened the pull request:
It implements sharded index writing and removes GBK stage in writing, relying on the read side sharding.

It exports a full genome BAM file of ~60GB in ~25min.

pgrosu commented:
Hi Ilia,

Looks really nice, but I don't have time to test it now - maybe a little later.  Just a few things in the meantime, in order to get the dependencies to the latest versions:

https://github.com/iliat/dataflow-java/blob/master/pom.xml#L72 -> s/1.1.0/1.4.0

https://github.com/iliat/dataflow-java/blob/master/pom.xml#L117 -> s/v1beta2-rev25-1.19.1/v1-rev56-1.21.0

https://github.com/iliat/dataflow-java/blob/master/pom.xml#L130 -> s/v1beta2-0.36/v1beta2-0.39

https://github.com/iliat/dataflow-java/blob/master/pom.xml#L194 -> s/1.128/2.1.0

https://github.com/iliat/dataflow-java/blob/master/pom.xml#L210 -> s/3.0.0-beta-1/3.0.0-beta-2

A re-test after these changes might not be a bad thing to perform, just to be sure everything passes.

Let me know what you think.

Thanks,
~p

deflaux commented on src/main/java/com/google/cloud/genomics/dataflow/utils/BreakFusionTransform.java:
```diff
@@ -0,0 +1,66 @@
+   static class DummyMapFn<T> extends DoFn<T, KV<T, Integer>> {  
+    @Override
+    public void processElement(DoFn<T, KV<T, Integer>>.ProcessContext c) throws Exception {
+      c.output( KV.of(c.element(), 42));
```
just a nit, perhaps move 42 to a DUMMY_VALUE constant?

add a link to https://cloud.google.com/dataflow/service/dataflow-service-desc#Optimization

iliat replied:
Great idea, DONE.

deflaux commented on src/main/java/com/google/cloud/genomics/dataflow/readers/bam/HeaderInfo.java:
```diff
@@ -0,0 +1,250 @@
+/*
```
This seems generally useful and not specific to dataflow.  Consider pushing it down to util-java in a subsequent PR.

iliat replied:
Yep, added https://github.com/googlegenomics/dataflow-java/issues/168

deflaux commented on src/main/java/com/google/cloud/genomics/dataflow/pipelines/CountReads.java:
```diff
@@ -176,16 +174,9 @@ private static boolean GCSURLExists(String url) {
-                .named(ReadReader.class.getSimpleName()));
+    final PCollection<Read> reads = p.begin()
+        .apply(Create.of(Collections.singletonList(pipelineOptions.getReadGroupSetId())))
+        .apply(new ReadGroupStreamer(auth, ShardBoundary.Requirement.STRICT, null, SexChromosomeFilter.INCLUDE_XY));
```
Nice update here.  I filed https://github.com/googlegenomics/readthedocs/issues/119

deflaux commented on src/main/java/com/google/cloud/genomics/dataflow/functions/KeyReadsFn.java:
```diff
@@ -52,15 +60,31 @@ public KeyReadsFn() {
     final Read read = c.element();
+    long pos = read.getAlignment().getPosition().getPosition();
+    if (lastPos > pos) {
+      LOG.info("Out of order read at pos "+ lastPos + " " + pos);
```
What is the implication of this event if it occurs?  A malformed input BAM file?  Or a bug in the pipeline?

This event is logged, but no action is taken.

pgrosu replied:
It's fine for the read positions in a BAM to be unsorted.  In fact you can add a followup merge/reduce step in the dataflow to have the reads sorted by position if you want.

iliat replied:
This is a leftover form the debugging I was doing - I wanted to confirm that when the data is read from the API via gRPC it does indeed come in coordinate order.
A bunch of optimizations rely on the fact that this is indeed the case, otherwise we have to reshard/reorder reads which is costly, especially in the case of full genome export (we end up materializing the whole thing several times in the process).
In the current implementation we in fact do not need to key the reads because we do not need to reshard them - we rely on reading side sharding and do not re-shard again for writing.
For the purposes of this function if it is used in other scenarios, where we do need resharding, the ordering does not strictly matter (as Paul says), so I removed the logging.
If some future implementation wants to rely on the ordering within a bundle we could implement a varian of KeyReadsFn that validates ordering and fails if it observes an out of order read but for now there is no concrete need for this.

deflaux commented:
@iliat this looks really awesome.  LGTM

Great speed up!  I assume mvn verify passed all integration tests.

Nice cleanup on the file names too.  Do you think some of this code is useful outside of the context of dataflow?  If so, some other time it would be nice to move it elsewhere.

pgrosu commented on src/main/java/com/google/cloud/genomics/dataflow/utils/BreakFusionTransform.java:
```diff
@@ -0,0 +1,66 @@
+import com.google.cloud.dataflow.sdk.values.PCollection;
+
+/*
+ * Breaks DataFlow fusion by doing GroubByKey/Ungroup that forces materialization of the data,
```
s/GroubByKey/GroupByKey

iliat replied:
Done

pgrosu commented on src/main/java/com/google/cloud/genomics/dataflow/utils/BreakFusionTransform.java:
```diff
@@ -0,0 +1,66 @@
+
+/*
+ * Breaks DataFlow fusion by doing GroubByKey/Ungroup that forces materialization of the data,
+ * thus preventing DataFlow form fusing steps before and after this transform.
```
s/form/from

pgrosu commented on src/main/java/com/google/cloud/genomics/dataflow/utils/BreakFusionTransform.java:
```diff
@@ -0,0 +1,66 @@
+ * thus preventing DataFlow form fusing steps before and after this transform.
+ * This is useful to insert in cases where a series of transforms deal with very small sets of data
+ * that act as descriptors of very heavy workloads in subsequent steps (e.g. a collection of file names
+ * where each file takes a long time to process).
```
nit, what would Dataflow define as long time before performing the fusion?

iliat replied:
Not sure I understand the question: I think you are asking, how long would DF wait executing a transform sequentially, before spinning up more workers (which I do not know), but in this case this is immaterial because if several steps are wrongly fused DF will not even attempt to spin up more workers because it "already decided this is going to be a single worker thing no matter what". So breaking fusion achieves exactly the freedom for DF to make a decision to spin up workers if it deems necessary.

pgrosu replied:
That makes sense.  Basically the reason I asked is that subsequently ([in lines 30-31](https://github.com/iliat/dataflow-java/blob/9fa7583736bd2422551a2ab0c4c0d016f22cfeba/src/main/java/com/google/cloud/genomics/dataflow/utils/BreakFusionTransform.java#L30-L31)) it's mentioned that:

 > In this case DataFlow might over-eagerly fuse steps dealing with small datasets with the "heavy" processing steps, which will result in heavy steps being executed on a single worker.

 I interpreted the word "with" as a necessary conjunction.  This indicated to me that the fusion is not only based on the small dataset size only - which would be nice to define as what is small - but rather that the fusion is also based on the heavy step in conjunction, which would indicate a job resubmit on less workers.

pgrosu commented on src/main/java/com/google/cloud/genomics/dataflow/pipelines/ShardedBAMWriting.java:
```diff
@@ -115,89 +141,24 @@ public static void main(String[] args) throws GeneralSecurityException, IOExcept
+     *    BAM_FILE_READ_SHARDING_POLICY = new ShardingPolicy() {
+     *     @Override
+     *     public boolean shardBigEnough(BAMShard shard) {
+     *       return shard.sizeInLoci() > 50000000;
```
Could `50000000` become a final variable or something that is tunable?  Basically having fixed values in the code might prove better by being defined via a variable.

iliat replied:
This is just a code example, the currently used value is already a constant, see ShardingPolicy.MAX_BYTES_PER_SHARD/MAX_LOCI_PER_SHARD constants.
The example is here to show what to do if you want your own decision making about shard sizes.

pgrosu commented on src/main/java/com/google/cloud/genomics/dataflow/pipelines/ShardedBAMWriting.java:
```diff
@@ -49,36 +37,59 @@
     void setBAMFilePath(String filePath);
     
+    @Description("An ID of the Google Genomics ReadGroupSets this " +
+        "pipeline is working with. This or BAMFilePath  must be set.")
```
s/BAMFilePath  must/BAMFilePath must

removes an extra space

iliat replied:
Done

pgrosu commented on src/main/java/com/google/cloud/genomics/dataflow/pipelines/ShardedBAMWriting.java:
```diff
@@ -208,7 +169,29 @@ public int compare(Contig o1, Contig o2) {
+              pipelineOptions.getReferences(), pipelineOptions.getBasesPerShard()));
+    }
+ 
+    LOG.info("Reading form the API with: "+ requests.size() + " shards");
```
s/"+/" +

an extra space

iliat replied:
Done

pgrosu commented on src/main/java/com/google/cloud/genomics/dataflow/pipelines/ShardedBAMWriting.java:
```diff
@@ -225,9 +208,4 @@ public Contig apply(String str) throws Exception {
           return Contig.parseContigsFromCommandLine(str).iterator().next();
```
s/str/contigStr

It will require the following update to this line from:

```Java
public Contig apply(String str) throws Exception {
```
to
```Java
public Contig apply(String contigStr) throws Exception {
```

This is to keep it consistent with [WriteReadsTransform](https://github.com/googlegenomics/dataflow-java/blob/de9258adc9bd8523d35058f43446b52e60332a27/src/main/java/com/google/cloud/genomics/dataflow/writers/WriteReadsTransform.java#L118)

iliat replied:
Done

pgrosu commented on src/main/java/com/google/cloud/genomics/dataflow/readers/bam/HeaderInfo.java:
```diff
@@ -0,0 +1,250 @@
+        this.firstRead.start >= shard.start && this.firstRead.start <= shard.end;
+  }
+  
+  public static HeaderInfo getHeaderFromApi(String  rgsId, OfflineAuth auth, Iterable<Contig> explicitlyRequestedContigs) 
```
s/String  rgsId/String rgsId

removal of an extra space

iliat replied:
Done

pgrosu commented on src/main/java/com/google/cloud/genomics/dataflow/utils/BreakFusionTransform.java:
```diff
@@ -0,0 +1,66 @@
+ *      .apply(new HeavyFileProcessingTransform())
+ *      .....
+ */
+public class BreakFusionTransform<T> extends PTransform<PCollection<T>,PCollection<T>> {
```
s/\<PCollection\<T\>,PCollection<T\>\>/\<PCollection\<T\>, PCollection\<T\>\>

Adds an extra space for easier readability

iliat replied:
Done

pgrosu commented on src/main/java/htsjdk/samtools/BAMShardIndexer.java:
```diff
@@ -0,0 +1,190 @@
+  // output written as binary, or (for debugging) as text
+  private final BinaryBAMShardIndexWriter outputWriter;
+
+  // content is built up from the input bam file using this
```
s/content/Content
or
s/content/The content

nit, s/this/a BamIndexBuilder

iliat replied:
This file and others in this folder are mostly copied as is from HTSJDK - I will add comments to that effect.
But I am going to largely ignore other comments about changes in this file since its not my intent to cleanup HTSJDK code that already works and in fact I'd rather preserve the code close to the original version, warts and all ;)

pgrosu replied:
No problem, but it would be nice to reference what was taken from other places.  Not everything I assume is from HTSJDK in my comments.  The `pom.xml` for instance would probably also need updating regarding the latest updates to packages, which might make the code run even faster.  

What you guys are doing is really important work to get high-throughput sequencing analysis to the next level, and proper documentation and consistent code will make reading and using the code more seamless and ubiquitous.  This way things would not need to be addressed later on in multiple places, which can become a headache.  I can take look again after the fixes.  I'm just helping you out man, as I've gone through things like this before :)

pgrosu commented on src/main/java/htsjdk/samtools/BAMShardIndexer.java:
```diff
@@ -0,0 +1,190 @@
+import java.io.OutputStream;
+
+public class BAMShardIndexer {
+  // output written as binary, or (for debugging) as text
```
s/ output/Output

or 

s/ output/The output is written

pgrosu commented on src/main/java/htsjdk/samtools/BAMShardIndexer.java:
```diff
@@ -0,0 +1,190 @@
+
+      private int currentReference = -1;
+
+      // information in meta data
```
s/ information in meta data/The metadata information is populated in a BAMIndexMetaData object

pgrosu commented on src/main/java/htsjdk/samtools/BAMShardIndexer.java:
```diff
@@ -0,0 +1,190 @@
+       */
+      public void processAlignment(final SAMRecord rec) {
+
+          // metadata
```
s/ metadata/Stores the metadata

pgrosu commented on src/main/java/htsjdk/samtools/BAMShardIndexer.java:
```diff
@@ -0,0 +1,190 @@
+              return; // do nothing for records without coordinates, but count them
+          }
+
+          // various checks
```
`various checks` is so general that I recommend it should be expanded for clarity.

pgrosu commented on src/main/java/htsjdk/samtools/BAMShardIndexer.java:
```diff
@@ -0,0 +1,190 @@
+          indexStats.recordMetaData(rec);
+
+          if (rec.getAlignmentStart() == SAMRecord.NO_ALIGNMENT_START) {
+              return; // do nothing for records without coordinates, but count them
```
Do you mean you count them via the `createIndex` function?  If so, then I recommend adding a bit more to this so folks make the connection in the code.

```Java
for (SAMRecord rec : reader) {
          if (++totalRecords % 1000000 == 0) {
              if (null != log) log.info(totalRecords + " reads processed ...");
          }
          indexer.processAlignment(rec);
      }
```

pgrosu commented on src/main/java/htsjdk/samtools/BAMShardIndexer.java:
```diff
@@ -0,0 +1,190 @@
+      }
+
+      /**
+       * reinitialize all data structures when the reference changes
```
s/reinitialize/Reinitialize

pgrosu commented on src/main/java/htsjdk/samtools/BAMShardIndexer.java:
```diff
@@ -0,0 +1,190 @@
+      void startNewReference(int reference) {
+          currentReference  = reference;
+          // I'm not crazy about recycling this object, but that is the way it was originally written and
+          // it helps keep track of no-coordinate read count (which shouldn't be stored in this class anyway).
```
So why not create a separate class to clean up the code?  You got my +1 vote :smile:  Less is always more in my book.

pgrosu commented on src/main/java/htsjdk/samtools/BAMShardIndexer.java:
```diff
@@ -0,0 +1,190 @@
+
+      int totalRecords = 0;
+
+      // create and write the content
```
s/create and write the content/Populates the index from the records in a BAM file.

pgrosu commented on src/main/java/htsjdk/samtools/BinaryBAMShardIndexWriter.java:
```diff
@@ -0,0 +1,158 @@
+      }
+
+      //final List<Chunk> chunks = content.getMetaData() == null ? null
+      //        : content.getMetaData().getMetaDataChunks();
```
If this code is not used, why not just remove it?

pgrosu commented on src/main/java/htsjdk/samtools/BinaryBAMShardIndexWriter.java:
```diff
@@ -0,0 +1,158 @@
+      final BAMIndexMetaData metaData = content.getMetaData();
+
+      codec.writeInt(size + ((metaData != null)? 1 : 0 ));
+      // codec.writeInt(size);
```
This code is commented out, why not just remove it?

pgrosu commented on src/main/java/htsjdk/samtools/BinaryBAMShardIndexWriter.java:
```diff
@@ -0,0 +1,158 @@
+          writeBin(bin);
+      }
+
+      // write metadata "bin" and chunks        
```
More expansion for clarity please in the description.

pgrosu commented on src/main/java/htsjdk/samtools/BinaryBAMShardIndexWriter.java:
```diff
@@ -0,0 +1,158 @@
+      if (metaData != null)
+          writeChunkMetaData(metaData);
+
+      // write linear index
```
I think I got that from the code, but not add why and the reason you need it.

pgrosu commented on src/main/java/htsjdk/samtools/BinaryBAMShardIndexWriter.java:
```diff
@@ -0,0 +1,158 @@
+      if (entries == null) {
+          return;
+      }
+      // since indexStart is usually 0, this is usually a no-op
```
So why not just put an if-else statement to make it more clear?

pgrosu commented on src/main/java/htsjdk/samtools/BinaryBAMShardIndexWriter.java:
```diff
@@ -0,0 +1,158 @@
+  /**
+   * Writes out the count of records without coordinates
+   *
+   * @param count
```
No description for `count`?  I know it's obvious, but usually all parameters have a description.

pgrosu commented on src/main/java/htsjdk/samtools/BinaryBAMShardIndexWriter.java:
```diff
@@ -0,0 +1,158 @@
+  }
+
+  /**
+   * Any necessary processing at the end of the file
```
The comment is too vague.  You are just flushing and closing the file, so a more specific description would be nice.

pgrosu commented on src/main/java/htsjdk/samtools/BinaryBAMShardIndexWriter.java:
```diff
@@ -0,0 +1,158 @@
+  }
+
+  /**
+   * Write the meta data represented by the chunkLists associated with bin MAX_BINS 37450
```
It would be nice to reference the following, so folks know where this is coming from:

https://github.com/samtools/htsjdk/blob/master/src/java/htsjdk/samtools/GenomicIndexUtil.java#L42=L45

pgrosu commented on src/main/java/htsjdk/samtools/BinaryBAMShardIndexWriter.java:
```diff
@@ -0,0 +1,158 @@
+  }
+
+  private void writeNullContent() {
+      codec.writeLong(0);  // 0 bins , 0 intv
```
Does `intv` mean interval?  It might be nice to expand on the comment here and why this is necessary, even if folks reading through the code will eventually figure it out.

pgrosu commented on src/main/java/htsjdk/samtools/BinaryBAMShardIndexWriter.java:
```diff
@@ -0,0 +1,158 @@
+      final LinearIndex linearIndex = content.getLinearIndex();
+      final long[] entries = linearIndex == null ? null : linearIndex.getIndexEntries();
+      final int indexStart = linearIndex == null ? 0 : linearIndex.getIndexStart();
+      final int n_intv = entries == null ? indexStart : entries.length + indexStart;
```
Does `n_intv` mean the nth interval?  If so, maybe something more descriptive like `nthIntervalFromStart` or `indexOffsetFromStart` might communicate quicker and easier.  Again we are talking about references so maybe tying that in into the variable name would allow for easier elucidation for the need of this variable.

jakeakopp commented on src/main/java/com/google/cloud/genomics/dataflow/functions/WriteBAIFn.java:
```diff
@@ -0,0 +1,141 @@
+ * The input is a reference name to process and the output is the name of the file written.
+ * A side output under the tag NO_COORD_READS_COUNT_TAG is a single value of the number of 
+ * no-coordinate reads in this shard. This is needed since the final index has to include a
+ * total number summed up form the shards.
```
s/form/from

iliat replied:
Done

jakeakopp commented on src/main/java/com/google/cloud/genomics/dataflow/pipelines/ShardedBAMWriting.java:
```diff
@@ -49,36 +37,59 @@
-    WriteReadsTransform.Options, GCSOutputOptions {
-    @Description("The Google Cloud Storage path to the BAM file to get reads data from")
+    WriteBAMTransform.Options, GCSOutputOptions {
+    @Description("The Google Cloud Storage path to the BAM file to get reads data from" + 
```
s/from/from. /

iliat replied:
Done

jakeakopp commented on src/main/java/com/google/cloud/genomics/dataflow/pipelines/ShardedBAMWriting.java:
```diff
@@ -49,36 +37,59 @@
-    @Description("The Google Cloud Storage path to the BAM file to get reads data from")
+    WriteBAMTransform.Options, GCSOutputOptions {
+    @Description("The Google Cloud Storage path to the BAM file to get reads data from" + 
+        "This or ReadGroupSetIds must be set")
```
ReadGroupSetId, here and line 87

iliat replied:
Done

jakeakopp commented:
@iliat Everything I can understand looks good ;-)

iliat commented:
@deflaux @pgrosu @jakeakopp Thanks for the review, I addressed most of the comments and will upload the version with these changes once I do a bit more validation beyond basic checks.

iliat commented:
@deflaux - I can finally update the PR with a version that:
1) Addresses comment above
2) Fixes index generation bug and fixes a long lurking bug GCSSeekableStream
3) Adds a BAMDiff tool I used to compare BAM files exported with API and this method to ensure no reads are missed (it ignores unmapped ones for now).

pgrosu commented:
Dude, there's some minor error in your JavaDoc - below is the link to the log for JDK8:

https://s3.amazonaws.com/archive.travis-ci.org/jobs/110838148/log.txt

Here's the start of the errors:

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:2.10.3:javadoc (default-cli) on project google-genomics-dataflow: An error has occurred in JavaDocs report generation:
[ERROR] Exit code: 1 - /home/travis/build/googlegenomics/dataflow-java/src/main/java/htsjdk/samtools/BAMShardIndexer.java:180: warning: no @param for log
[ERROR] public static void createIndex(SamReader reader, File output, Log log) {
[ERROR] ^
[ERROR] /home/travis/build/googlegenomics/dataflow-java/src/main/java/htsjdk/samtools/BAMShardIndexer.java:10: error: unexpected text
[ERROR] * @see https://github.com/samtools/htsjdk/blob/master/src/java/htsjdk/samtools/BAMIndexer.java
[ERROR] ^
[ERROR] /home/travis/build/googlegenomics/dataflow-java/src/main/java/htsjdk/samtools/BAMShardIndexer.java:14: warning - Tag @see:illegal character: "58" in "https://github.com/samtools/htsjdk/blob/master/src/java/htsjdk/samtools/BAMIndexer.java
[ERROR] and modified to support sharded index writing, where index for each reference is generated
[ERROR] separately and then the index shards are combined."
[ERROR] /home/travis/build/googlegenomics/dataflow-java/src/main/java/htsjdk/samtools/BAMShardIndexer.java:14: warning - Tag @see:illegal character: "47" in "https://github.com/samtools/htsjdk/blob/master/src/java/htsjdk/samtools/BAMIndexer.java
[ERROR] and modified to support sharded index writing, where index for each reference is generated
[ERROR] separately and then the index shards are combined."
[ERROR] /home/travis/build/googlegenomics/dataflow-java/src/main/java/htsjdk/samtools/BAMShardIndexer.java:14: warning - Tag @see:illegal character: "47" in "https://github.com/samtools/htsjdk/blob/master/src/java/htsjdk/samtools/BAMIndexer.java
[ERROR] and modified to support sharded index writing, where index for each reference is generated
[ERROR] separately and then the index shards are combined."
[ERROR] /home/travis/build/googlegenomics/dataflow-java/src/main/java/htsjdk/samtools
```

iliat commented:
@pgrosu Yeah, on it :)

pgrosu commented:
Ah, cool man :)

pgrosu commented on src/main/java/com/google/cloud/genomics/dataflow/utils/BAMDiff.java:
```diff
@@ -0,0 +1,457 @@
+    ret = compareReadGroups(h1, h2) && ret;
+    ret = compareProgramRecords(h1, h2) && ret;
+    return ret;
+}
```
Way too cryptic dude!  Try the following instead free of charge :)

```Java
private boolean compareHeaders(SAMFileHeader h1, SAMFileHeader h2) throws Exception {
    
    if ( !compareSequenceDictionaries(h1, h2) ) {
      
      return false;

    } else if (

      compareValues(h1.getCreator(), h2.getCreator(), "File creator") &&
      compareValues(h1.getAttribute("SO"), h2.getAttribute("SO"), "Sort order") &&
      compareReadGroups(h1, h2) &&
      compareProgramRecords(h1, h2) ) {

        if ( !options.ignoreFileFormatVersion ) {
          return compareValues(h1.getVersion(), h2.getVersion(), "File format version");
        } else {
          return true;
        }

    } else {
      return false;
    }  
}
```

pgrosu commented on src/main/java/com/google/cloud/genomics/dataflow/utils/BAMDiff.java:
```diff
@@ -0,0 +1,457 @@
+      }
+    }
+  
+    return ret;
```
How about turning this from O(N^2) to O(N) via:

```Java
String pr1IDs = new String();
String pr2IDs = new String();

for (SAMProgramRecord pr1 : l1) {
	pr1IDs = new String( pr1IDs + pr1.getId() );
}

for (SAMProgramRecord pr2 : l2) {
	pr2IDs = new String( pr2IDs + pr2.getId() );
}

return pr1IDs.equals(pr2IDs);
```

pgrosu commented on src/main/java/com/google/cloud/genomics/dataflow/utils/BAMDiff.java:
```diff
@@ -0,0 +1,457 @@
+    for (int i = 0; i < l1.size(); ++i) {
+        ret = compareReadGroup(l1.get(i), l2.get(i)) && ret;
+    }
+    return ret;
```
Maybe something simpler such as:

```Java
for (int i = 0; i < l1.size(); ++i) {
  if ( !compareReadGroup(l1.get(i), l2.get(i)) ) {
    return false;
  }
}
return true;
```

pgrosu commented on src/main/java/com/google/cloud/genomics/dataflow/utils/BAMDiff.java:
```diff
@@ -0,0 +1,457 @@
+                attribute + " for read group " + samReadGroupRecord1.getReadGroupId()) && ret;
+    }
+    return ret;
+}
```
How about something simpler like:

```Java
private boolean compareReadGroup( final SAMReadGroupRecord samReadGroupRecord1, 
                                  final SAMReadGroupRecord samReadGroupRecord2) throws Exception {

  if (
       compareValues(samReadGroupRecord1.getReadGroupId(), samReadGroupRecord2.getReadGroupId(),
              "Read Group ID") &&

       compareValues(samReadGroupRecord1.getSample(), samReadGroupRecord2.getSample(),
              "Sample for read group " + samReadGroupRecord1.getReadGroupId()) &&
          
       compareValues(samReadGroupRecord1.getLibrary(), samReadGroupRecord2.getLibrary(),
              "Library for read group " + samReadGroupRecord1.getReadGroupId())) {

          final String[] attributes = {"DS", "PU", "PI", "CN", "DT", "PL"};
          
          for (final String attribute : attributes) {

            String a1 = samReadGroupRecord1.getAttribute(attribute);
            String a2 = samReadGroupRecord2.getAttribute(attribute);
            if (options.ignoreNullVsZeroPI && attribute.equals("PI")) {
              if (a1 == null) {
                a1 = "0";
              }
              if (a2 == null) {
                a2 = "0";
              }
            }
            
            if( !compareValues(a1, a2,
                    attribute + " for read group " + samReadGroupRecord1.getReadGroupId())) {

                return false;
              }
          }

       } else {
          return true;
       }  
}
```

pgrosu commented on src/main/java/com/google/cloud/genomics/dataflow/utils/BAMDiff.java:
```diff
@@ -0,0 +1,457 @@
+              sequenceRecord1.getSequenceName()) && ret;
+    }
+    return ret;
+}
```
How about the following to simplify:

```Java
private boolean compareSequenceRecord( final SAMSequenceRecord sequenceRecord1, 
                                       final SAMSequenceRecord sequenceRecord2, final int which) throws Exception {

   if ( !sequenceRecord1.getSequenceName().equals(sequenceRecord2.getSequenceName())) {
        reportDifference(sequenceRecord1.getSequenceName(), sequenceRecord2.getSequenceName(),
                "Name of sequence record " + which);
        return false;
   }

   if ( compareValues(sequenceRecord1.getSequenceLength(), sequenceRecord2.getSequenceLength(), "Length of sequence " +
            sequenceRecord1.getSequenceName())) {

    if ( !options.ignoreSequenceProperties &&
        
         compareValues(sequenceRecord1.getSpecies(), sequenceRecord2.getSpecies(), "Species of sequence " +
              sequenceRecord1.getSequenceName()) && 

          compareValues(sequenceRecord1.getAssembly(), sequenceRecord2.getAssembly(), "Assembly of sequence " +
              sequenceRecord1.getSequenceName()) &&

          compareValues(sequenceRecord1.getAttribute("M5"), sequenceRecord2.getAttribute("M5"), "MD5 of sequence " +
              sequenceRecord1.getSequenceName()) &&

          compareValues(sequenceRecord1.getAttribute("UR"), sequenceRecord2.getAttribute("UR"), "URI of sequence " +
              sequenceRecord1.getSequenceName()) ) {
      
      return true;
    } else {
      return false;
    }
  }
  return false;
}
```

pgrosu commented on src/main/java/com/google/cloud/genomics/dataflow/utils/BAMDiff.java:
```diff
@@ -0,0 +1,457 @@
+      }
+    }
+    return ret;
+}
```
How about the following to simplify:

```Java
private boolean compareSequenceDictionaries(final SAMFileHeader h1, final SAMFileHeader h2) throws Exception {

    final List<SAMSequenceRecord> s1 = h1.getSequenceDictionary().getSequences();
    final List<SAMSequenceRecord> s2 = h2.getSequenceDictionary().getSequences();
  
    if (referencesToProcess == null && s1.size() != s2.size()) {
        reportDifference(s1.size(), s2.size(), "Length of sequence dictionaries");
        return false;
    }
  
    if (referencesToProcess == null) {
      LOG.info("Comparing all sequences in the headers");
      for (int i = 0; i < s1.size(); ++i) {
        LOG.info("Comparing reference at index " + i);
        SAMSequenceRecord sr1 = s1.get(i);
        SAMSequenceRecord sr2 = options.ignoreSequenceOrder ?
            h2.getSequenceDictionary().getSequence(sr1.getSequenceName()) : 
            s2.get(i);
        if (sr2 == null) {
          error("Failed to find sequence " + sr1.getSequenceName() + " in " + BAMFile2);
        }
        if( !compareSequenceRecord(sr1, sr2, i + 1) ) {
          return false;
        }
      }
    } else {
      LOG.info("Comparing specified sequences in the headers");
      for (String r : referencesToProcess) {
        LOG.info("Comparing reference " + r);
        if( !compareSequenceRecord(h1.getSequenceDictionary().getSequence(r),
            h2.getSequenceDictionary().getSequence(r), -1) ) { 
          return false;
        }
      }
    }
    return true;
}
```

pgrosu commented on src/main/java/com/google/cloud/genomics/dataflow/utils/BAMDiff.java:
```diff
@@ -0,0 +1,457 @@
+        }
+        reportDifference(v1, v2, label);
+        return false;
+    }
```
How about the following for just this portion:

```Java
    if ( v1 == null && v2 == null ) {
      return true;
    } else {
      reportDifference(v1, v2, label);
      return false;
    }
```

pgrosu commented on src/main/java/com/google/cloud/genomics/dataflow/utils/BAMDiff.java:
```diff
@@ -0,0 +1,457 @@
+    }
+    reportDifference(o1.toString(), o2.toString(), label);
+}
+}
```
Probably all the methods need just one indent as the curly brackets at the end align.

iliat commented:
PTAL @deflaux

pgrosu commented:
Wow Ilia!  It's so beautiful, that it brings me tears of joy :)

![cat_joy](https://cloud.githubusercontent.com/assets/6555937/13406223/52cbc010-def0-11e5-86c5-1b1d2ae5ab37.png)

deflaux commented:
LGTM @iliat Again, really nice work here.

Please file issues for stuff to be done in future PRs.  Thanks!!!
